src/qinfer/domains.py

```diff
@@ -145,6 +148,149 @@ def in_domain(self, points):
         """
         pass
 
+class ProductDomain(Domain):
+    """
+    A domain made from the cartesian product of other domains.
+
+    :param Domain domains: ``Domain`` instances as separate arguments, 
+        or as a singe list of ``Domain`` instances.
+    """
+    def __init__(self, *domains):
+        
+        if len(domains) == 1:
+            try:
+                self._domains = list(domains[0])
+            except:
+                self._domains = domains
+        else:
+            self._domains = domains
+        
+        self._domains = domains
+        self._dtypes = [domain.example_point.dtype for domain in self._domains]
+        self._example_point = join_struct_arrays(
+            [np.array(domain.example_point) for domain in self._domains]
+        )
+        self._dtype = self._example_point.dtype
+
+    @property
+    def is_continuous(self):
+        """
+        Whether or not the domain has an uncountable number of values.
+
+        :type: `bool`
+        """
+        return any([domain.is_continuous for domain in self._domains])
+
+    @property
+    def is_finite(self):
+        """
+        Whether or not the domain contains a finite number of points.
+
+        :type: `bool`
+        """
+        return all([domain.is_finite for domain in self._domains])
+
+    @property
+    def dtype(self):
+        """
+        The numpy dtype of a single element of the domain.
+
+        :type: `np.dtype`
+        """
+        return self._dtype
+
+    @property
+    def n_members(self):
+        """
+        Returns the number of members in the domain if it
+        `is_finite`, otherwise, returns `np.inf`.
+
+        :type: ``int`` or ``np.inf``
+        """
+        if self.is_finite:
+            return reduce(mul, [domain.n_members for domain in self._domains], 1)
```

Comment:
This line is raising a Py3 compatibility warning; should be able to fix by importing ``reduce`` from ``functools``.